File: code/97-Interleaving String/97-Interleaving String.cpp
// Given strings s1, s2, and s3, find whether s3 is formed by an interleaving of s1 and s2.

// An interleaving of two strings s and t is a configuration where they are divided into non-empty substrings such that:

// s = s1 + s2 + ... + sn
// t = t1 + t2 + ... + tm
// |n - m| <= 1
// The interleaving is s1 + t1 + s2 + t2 + s3 + t3 + ... or t1 + s1 + t2 + s2 + t3 + s3 + ...
// Note: a + b is the concatenation of strings a and b.

 

// Example 1:


// Input: s1 = "aabcc", s2 = "dbbca", s3 = "aadbbcbcac"
// Output: true
// Example 2:

// Input: s1 = "aabcc", s2 = "dbbca", s3 = "aadbbbaccc"
// Output: false
// Example 3:

// Input: s1 = "", s2 = "", s3 = ""
// Output: true
 

// Constraints:

// 0 <= s1.length, s2.length <= 100
// 0 <= s3.length <= 200
// s1, s2, and s3 consist of lowercase English letters.
 

// Follow up: Could you solve it using only O(s2.length) additional memory space?

// class Solution {
// public:
//     bool isInterleave(string s1, string s2, string s3) {
//         int n1 = s1.size(), n2 = s2.size(), n3 = s3.size();
//         if (n1 + n2 != n3) {
//             return false;
//         }
//         vector<vector<bool>> boolean(n2 + 1, vector<bool>(n1 + 1, false));
//         boolean[0][0] = true;
//         for (int i = 0; i <= n1; i ++) {
//             for (int j = 0; j <= n2; j ++) {
//                 int p = i + j - 1;
//                 if (i > 0) {
//                     boolean[i][j] = boolean[i][j] || (boolean[i - 1][j] && s1[i - 1] == s3[p]);
//                 }
//                 if (j > 0) {
//                     boolean[i][j] = boolean[i][j] || (boolean[i][j - 1] && s2[j - 1] == s3[p]);
//                 }
//             }
//         }
//         return boolean[n1][n2];
//     }
// };


class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        auto f = vector <int> (s2.size() + 1, false);

        int n = s1.size(), m = s2.size(), t = s3.size();

        if (n + m != t) {
            return false;
        }

        f[0] = true;
        for (int i = 0; i <= n; ++i) {
            for (int j = 0; j <= m; ++j) {
                int p = i + j - 1;
                if (i > 0) {
                    f[j] &= (s1[i - 1] == s3[p]);
                }
                if (j > 0) {
                    f[j] |= (f[j - 1] && s2[j - 1] == s3[p]);
                }
            }
        }

        return f[m];
    }
};

File: code/97-Interleaving String/97-Interleaving String.py
# -*- coding: utf-8 -*
"""
LeetCode 97 - Interleaving String

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
"""
r"""
@author: Kaka Chen
@license: (C) Copyright Personal Uncommercial Limited.
@contact: kakachen1990@gmail.com
@file: 97-Interleaving String.py
@time: 2020/7/18 11:15 上午
@desc:
Given s1, s2, s3, find whether s3 is formed by the interleaving of s1 and s2.

Example 1:

Input: s1 = "aabcc", s2 = "dbbca", s3 = "aadbbcbcac"
Output: true
Example 2:

Input: s1 = "aabcc", s2 = "dbbca", s3 = "aadbbbaccc"
Output: false
"""


class Solution(object):
    def isInterleave(self, s1, s2, s3):
        """
        :type s1: str
        :type s2: str
        :type s3: str
        :rtype: bool
        """
        n1, n2, n3 = len(s1), len(s2), len(s3)
        if n1 + n2 != n3:
            return False
        boolean = [[False] * (n2 + 1) for _ in range(n1 + 1)]
        boolean[0][0] = True
        for i in range(n1 + 1):
            for j in range(n2 + 1):
                p = i + j - 1
                if i > 0:
                    boolean[i][j] = boolean[i][j] or (boolean[i - 1][j] and s1[i - 1] == s3[p])
                if j > 0:
                    boolean[i][j] = boolean[i][j] or (boolean[i][j - 1] and s2[j - 1] == s3[p])
        return boolean[n1][n2]


if __name__ == '__main__':
    s = Solution()
    # s1 = "aabcc"
    # s2 = "dbbca"
    # s3 = "aadbbcbcac"
    # print(s.isInterleave(s1, s2, s3))
    # s1 = "aabcc"
    # s2 = "dbbca"
    # s3 = "aadbbbaccc"
    # print(s.isInterleave(s1, s2, s3))
    s1 = "db"
    s2 = "b"
    s3 = "cbb"
    print(s.isInterleave(s1, s2, s3))
